crates/transpiler/src/transpile_layout.rs

```diff
@@ -0,0 +1,1041 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use hashbrown::HashMap;
+
+use qiskit_circuit::bit::ShareableQubit;
+use qiskit_circuit::nlayout::{NLayout, PhysicalQubit, VirtualQubit};
+
+// TODO: Conditionally compile these imports for Python builds
+use pyo3::intern;
+use pyo3::prelude::*;
+use pyo3::types::PyDict;
+use qiskit_circuit::imports::LAYOUT;
+use qiskit_circuit::imports::TRANSPILE_LAYOUT;
+
+/// The "layout" caused by transpilation
+///
+/// In general Qiskit's transpiler is unitary-preserving up to the initial layout
+/// and routing permutations. The initial layout permutation is caused by
+/// setting and applying the initial layout (the mapping from virtual circuit
+/// qubits to physical qubits on the target) and the routing permtuations are
+/// caused by swap gate insertion or permutation ellision prior to the initial
+/// layout. This struct tracks these details and provide an interface to reason
+/// about these permutations.
+pub struct TranspileLayout {
+    /// The initial layout which is mapping the virtual qubits in the input circuit to the
+    /// transpiler to the physical qubits used on the transpilation target
+    initial_layout: NLayout,
```

Comment:
Based on a comment I've seen somewhere, do we want to make `initial_layout` into an `Option`? In particular, we would then be able to easily determine whether the layout is already applied.

Terminology-wise, would it make sense to call this `layout` rather than `initial_layout`?

Reply:
Sure I think making this optional makes sense, as for the naming I know that you and @jakelishman have a plan on refactoring the way we reason about layout in the transpiler. I went with initial layout because that's what it was called in Python, but I'm fine calling it `layout` if that's preferable I don't really have much preference as long as it's clear to everyone.

Reply:
For naming: imo `initial_layout` is fine.  Since our plan is to get rid of `final_layout` as something that's _explicitly_ stored, I agree that `initial_` isn't needed for disambiguation, but in the prospective new object, there's still going to be "initial relabelling" and "final implicit [stuff]", so "initial" is still fine, and makes things clear.